File: Linear/Dynamic/Linked List/Circular Linked List/01_Inserting/Given_Position/Inserting.java
import java.util.Scanner;

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class InsertingGivenPosition {
    public static Node insertAtGivenPosition(Node last, int pos, int x) {
        Node newNode = new Node(x);

        if (last == null) {
            newNode.next = newNode;
            return newNode;
        }

        if (pos == 1) {
            newNode.next = last.next;
            last.next = newNode;
    
            return last;
        }

        Node curr = last.next;

        for (int i = 1; i < pos - 1 && curr != last; i++) {
            curr = curr.next;
        }

        newNode.next = curr.next;
        curr.next = newNode;

        if (curr == last)
        last = newNode;

        return last;
    }

    public static void printList(Node last) {
        if (last == null)
        return;

        Node head = last.next;
        Node temp = head;

        do {
            System.out.print(temp.data);
            temp = temp.next;
            if (temp != head)
            System.out.print(" -> ");
        } while (temp != head);

        System.out.println();
    }

    public static void main(String[] args) {
        // Create circular linked list: 2 -> 3 -> 4
        Node first = new Node(2);
        first.next = new Node(3);
        first.next.next = new Node(4);

        Node last = first.next.next;
        last.next = first;

        System.out.println("Linked List before insertion");
        printList(last);

        Scanner sc = new Scanner(System.in);
        System.out.println("Enter a position: ");
        int x = 5, pos = sc.nextInt();
        last = insertAtGivenPosition(last, pos, x);
        sc.close();

        System.out.println("Linked List after insertion");
        printList(last);
    }
}

File: Linear/Dynamic/Linked List/Circular Linked List/01_Inserting/Beginning/Inserting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insertAtBeginning(last, x):
    newNode = Node(x)

    if last is None:
        newNode.next = newNode
        return newNode
    
    newNode.next = last.next
    last.next = newNode

    return last

def printList(last):
    if last is None:
        return
    
    head = last.next
    temp = head

    while True:
        print(temp.data, end="")
        temp = temp.next
        if temp is not head:
            print(" -> ", end="")
        else:
            break
    
    print()

# Create circular linked list: 2 -> 3 -> 4
first = Node(2)
first.next = Node(3)
first.next.next = Node(4)

last = first.next.next
last.next = first

print("Linked List before insertion\n")
printList(last)

x = 5
last = insertAtBeginning(last, x)

print("Linked List after insertion\n")
printList(last)